Fix -Wimplicit-fallthrough warnings from GCC 8

akien-mga opened the pull request:
Adds `FALLTHROUGH` macro to specify when a fallthrough is intentional.
Can be replaced by `[[fallthrough]]` if/when we switch to C++17.

The warning is now enabled by default for GCC on `extra` warnings level
(part of GCC's `-Wextra`). It's not enabled in Clang's `-Wextra` yet,
but we could enable it manually once we switch to C++11. There's no
equivalent feature in MSVC for now.

Fixes #26135.

*Edited comment to match current implementation.*

akien-mga commented:
CC @marxin

akien-mga commented:
@marxin I've spent a while trying to make the default `-Wimplicit-fallthrough=3` work, to avoid having to specify a different level manually and further clutter the build command, but to no avail.

Either the documentation is wrong (see the regexes on https://gcc.gnu.org/onlinedocs/gcc-8.3.0/gcc/Warning-Options.html#index-Wimplicit-fallthrough), or there's a bug in GCC 8.3.0.

I've tried both `// -fallthrough` and `// intentional fallthrough`, which should match respectively the ` -fallthrough` and `[ \t.!]*([Ee]lse,? |[Ii]ntentional(ly)? )?
fall(s | |-)?thr(ough|u)[ \t.!]*(-[^\n\r]*)?` regexes from the documentation for `=3`, but I still got various false positives.

akien-mga commented:
> not included in Clang's
-Wextra for now, which is best as they don't support fallthrough comments
as annotations

If Clang does add `-Wimplicit-fallthrough` to their `-Wextra` in the future, we'll have two solutions:
- Implement our own `GODOT_FALLTHROUGH` macro that conditionally expands to `// fallthrough` or `__attribute__ ((fallthrough));` for GCC and `[[clang:fallthrough]]` for Clang (yay for more compiler-specific annotations!)
- Move to C++17 and use the new `[[fallthrough]];` attribute.

neikeq commented:
I think it would be better to have a `FALLTHROUGH` macro that expands to `[[fallthrough]]` with C++17, `__attribute__((fallthrough))` with GCC 7+, or a comment. Would a comment inside the macro work to get rid of the warning?

akien-mga commented:
> I think it would be better to have a FALLTHROUGH macro that expands to [[fallthrough]] with C++17, __attribute__((fallthrough)) with GCC 7+, or a comment. Would a comment inside the macro work to get rid of the warning?

The comment is an alternative way to satisfy GCC, so if we're adding a macro using `__attribute__((fallthrough))`, we wouldn't need the comment at all.

But yeah I guess I could make a `FALLTHROUGH` macro to support GCC and Clang's specific annotations, and if we ever move to C++17, we can replace it by `[[fallthrough]]`.

marxin commented:
> But yeah I guess I could make a `FALLTHROUGH` macro to support GCC and Clang's specific annotations, and if we ever move to C++17, we can replace it by `[[fallthrough]]`.

I like the idea of defining a macro. I'll take a look tomorrow whether it works with GCC.
Btw. are you planning to switch to C++17 soonish?

marxin commented:
> I've tried both `// -fallthrough` and `// intentional fallthrough`, which should match respectively the ` -fallthrough` and `[ \t.!]*([Ee]lse,? |[Ii]ntentional(ly)? )? fall(s | |-)?thr(ough|u)[ \t.!]*(-[^\n\r]*)?` regexes from the documentation for `=3`, but I still got various false positives.

About this, can you please send me a patch where you see a false positives? I can eventually report that to ```GCC``` bugzilla.

akien-mga commented:
> I like the idea of defining a macro. I'll take a look tomorrow whether it works with GCC.

I'll give this a try now, Boost has a `BOOST_FALLTHROUGH` macro for that which we can probably reuse. I'm not sure if Clang's `[[clang:fallthrough]]` works before C++11 though.

> Btw. are you planning to switch to C++17 soonish?

Not for now. We plan to switch to C++11 in coming weeks though, and maybe C++14 if we find enough stuff in the standard that we want to use.

> About this, can you please send me a patch where you see a false positives? I can eventually report that to GCC bugzilla.

Just bump to `-Wimplicit-fallthrough=3` in this PR and you should get warnings even though the fallthrough are properly marked as intentional as per the regex.

I'll see if I can make a minimum reproduction outside Godot's codebase.

marxin commented:
> Just bump to `-Wimplicit-fallthrough=3` in this PR and you should get warnings even though the fallthrough are properly marked as intentional as per the regex.

Good.

> I'll see if I can make a minimum reproduction outside Godot's codebase.

Won't be necessary then.

akien-mga commented:
Boost's definition:
```
config/detail/suffix.hpp
962-//
963-// Helper macro BOOST_FALLTHROUGH
964-// Fallback definition of BOOST_FALLTHROUGH macro used to mark intended
965-// fall-through between case labels in a switch statement. We use a definition
966-// that requires a semicolon after it to avoid at least one type of misuse even
967-// on unsupported compilers.
968-//
969-#ifndef BOOST_FALLTHROUGH
970:#  define BOOST_FALLTHROUGH ((void)0)
971-#endif

config/compiler/gcc.hpp
306-#if __GNUC__ >= 7
307:#  define BOOST_FALLTHROUGH __attribute__((fallthrough))
308-#endif

config/compiler/xlcpp.hpp
68-//
69-// The BOOST_FALLTHROUGH macro can be used to annotate implicit fall-through
70-// between switch labels.
71-//
72-#if __cplusplus >= 201103L && defined(__has_warning)
73-#  if __has_feature(cxx_attributes) && __has_warning("-Wimplicit-fallthrough")
74:#    define BOOST_FALLTHROUGH [[clang::fallthrough]]
75-#  endif
76-#endif

config/compiler/clang.hpp
111-//
112-// The BOOST_FALLTHROUGH macro can be used to annotate implicit fall-through
113-// between switch labels.
114-//
115-#if __cplusplus >= 201103L && defined(__has_warning)
116-#  if __has_feature(cxx_attributes) && __has_warning("-Wimplicit-fallthrough")
117:#    define BOOST_FALLTHROUGH [[clang::fallthrough]]
118-#  endif
119-#endif
```

akien-mga commented:
> Just bump to -Wimplicit-fallthrough=3 in this PR and you should get warnings even though the fallthrough are properly marked as intentional as per the regex.

@marxin I pushed this test in a specific branch of my fork, as I'm going to rebase this PR to use a macro: https://github.com/akien-mga/godot/tree/Wimplicit-fallthrough-test-level-3
Travis CI running on it, if all goes "well" it should fail: https://travis-ci.org/akien-mga/godot/builds/516124405

akien-mga committed:
SCons: add `methods.using_clang` to check used compiler

Also rename `use_gcc` to `using_gcc` to make it clear that it returns
a config but does not alter it.

akien-mga commented:
I wrote a macro inspired from the Boost one:
```cpp
/** Hint for compilers that this fallthrough in a switch is intentional.
 *  Can be replaced by [[fallthrough]] annotation if we move to C++17.
 *  Including conditional support for it for people who set -std=c++17
 *  themselves.
 *  Requires a trailing semicolon when used.
 */
#if __cplusplus >= 201703L
#define FALLTHROUGH [[fallthrough]]
#elif defined(__GNUC__) && __GNUC__ >= 7
#define FALLTHROUGH __attribute__((fallthrough))
#elif defined(__llvm__) && __cplusplus >= 201103L && __has_feature(cxx_attributes) \
		&& defined(__has_warning) && __has_warning("-Wimplicit-fallthrough")
#define FALLTHROUGH [[clang::fallthrough]]
#else
#define FALLTHROUGH
#endif
```

The problem is, as I feared, that clang-format makes a mess with it:

```cpp
			case TreeItem::CELL_MODE_CHECK: {

				int check_icon_h = cache.checked->get_height();
				if (height < check_icon_h)
					height = check_icon_h;
			} FALLTHROUGH;
			case TreeItem::CELL_MODE_STRING:
```
becomes:
```cpp
			case TreeItem::CELL_MODE_CHECK: {

				int check_icon_h = cache.checked->get_height();
				if (height < check_icon_h)
					height = check_icon_h;
			}
				FALLTHROUGH;
			case TreeItem::CELL_MODE_STRING:
```

Not sure why and how to work it around :/

akien-mga commented:
BTW I've confirmed that the macro works fine for GCC with default `-Wextra` (so `-Wimplicit-fallthrough=3`). I'll see if it would work when putting the macro inside the braces:
```cpp
			case TreeItem::CELL_MODE_CHECK: {

				int check_icon_h = cache.checked->get_height();
				if (height < check_icon_h)
					height = check_icon_h;
					FALLTHROUGH;
			}
			case TreeItem::CELL_MODE_STRING:
```
which should satisfy clang-format, even if it doesn't match the style we use otherwise for `} break;` in many places.

akien-mga committed:
Fix -Wimplicit-fallthrough warnings from GCC 8

Adds `FALLTHROUGH` macro to specify when a fallthrough is intentional.
Can be replaced by `[[fallthrough]]` if/when we switch to C++17.

The warning is now enabled by default for GCC on `extra` warnings level
(part of GCC's `-Wextra`). It's not enabled in Clang's `-Wextra` yet,
but we could enable it manually once we switch to C++11. There's no
equivalent feature in MSVC for now.

Fixes #26135.

akien-mga commented:
Rebased to add a `FALLTHROUGH` macro, it should work fine now with GCC 8's default `-Wextra`. Since the GCC attribute is for GCC 7+, I assume that `-Wimplicit-fallthrough` is not defined in GCC < 7's `-Wextra`, I guess the CI build will confirm that.

We can enable the warning explicitly for Clang once we move to C++11.

marxin commented:
Great. So with the usage of the macro, there are no issues with false positives, right?

akien-mga commented:
> Great. So with the usage of the macro, there are no issues with false positives, right?

Yeah, `__attribute__((fallthrough))__` seems more reliable than the documented comment regexes. Would still be worth reviewing that and confirm if there's a bug in GCC or its documentation.

I wonder if the regex was choking on situations like this:
```cpp
case 1: {
	stuff();
}; // fallthrough
case 2:
```
i.e. being misled by the closing brace and/or the (unnecessary) semicolon.

The format documented in the man page is:
```cpp
case 1:
	stuff();
	// fallthrough
case 2:
```

marxin commented:
I'll take a look tomorrow.

akien-mga commented:
@marxin Minimal reproducer for the GCC 8.3.0 false positive:
```cpp
#include <cstdio>

int main() {
	int i = 0;
	switch(i) {
		case 0:
			printf("%d\n", i);
			// intentional fallthrough to case 1
		case 1:
			printf("%d\n", i + 1);
			// intentional fallthrough
		case 2:
			printf("%d\n", i + 2);
			break;
	}
	return 0;
}
```
```
$ g++ --version
g++ (Mageia 8.3.1-0.20190329.1.mga7) 8.3.1 20190329

$ g++ -Wimplicit-fallthrough=3 main.cpp 
main.cpp: In function ‘int main()’:
main.cpp:7:10: warning: this statement may fall through [-Wimplicit-fallthrough=]
    printf("%d\n", i);
    ~~~~~~^~~~~~~~~~~
main.cpp:9:3: note: here
   case 1:
   ^~~~
```

`case 0` gives a warning because there is additional text after `fallthrough` in the comment. `case 1` works.

So the regex used is not the one documented, as that one should support extra text after "fallthr(ough|u)":
```
[ \t.!]*([Ee]lse,? |[Ii]ntentional(ly)? )?fall(s | |-)?thr(ough|u)[ \t.!]*(-[^\n\r]*)?
```
https://regex101.com/r/PjJLem/1

akien-mga commented:
> So the regex used is not the one documented, as that one should support extra text after "fallthr(ough|u)":
> ```
> [ \t.!]*([Ee]lse,? |[Ii]ntentional(ly)? )?fall(s | |-)?thr(ough|u)[ \t.!]*(-[^\n\r]*)?
> ```
> https://regex101.com/r/PjJLem/1

Actually reading the docs and regex again, it states "The **whole body** of the C or C++ style comment should match the given regular expressions listed below." (emphasis mine).
So I guess it's behaving as expected, `[ \t.!]*(-[^\n\r]*)?` does not match arbitrary text after "fallthrough". IMO it's needlessly strict and hard to get right, but it's not a bug.

akien-mga commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -910,7 +910,7 @@ void ScriptTextEditor::_edit_option(int p_op) {
 				ScriptEditor::get_singleton()->get_debugger()->set_breakpoint(script->get_path(), line + 1, dobreak);
 			}
-		}
+		} break;
```
Note: I'm changing the logic here, but the previous code did not seem like this fallthrough was intentional. I haven't tested (yet) to confirm.

akien-mga commented on scene/2d/collision_object_2d.cpp:
```diff
@@ -56,15 +57,15 @@ void CollisionObject2D::_notification(int p_what) {
 
 			//get space
-		}
+		} break;
```
Here too, I added `break:`s as the fallthrough did not seem intentional, but if it is, this change might cause bugs.

marxin commented:
> So I guess it's behaving as expected, `[ \t.!]*(-[^\n\r]*)?` does not match arbitrary text after "fallthrough". IMO it's needlessly strict and hard to get right, but it's not a bug.

I must agree with you, what's strange is that one can provide arbitrary text at the end, but only if dash (`-`) is used. Like here:
https://regex101.com/r/uH0cgh/2

I can discuss that with the author of the warning.

marxin commented:
> I can discuss that with the author of the warning.

I did so and I was told the format is intended.

akien-mga commented:
IMO the formats used in those warnings levels are quite project-specific, so it looks like the author upstreamed their own project's convention together with the warning... but anyway, the `__attribute__(fallthrough)` is the most reliable one, so what we merge is the right approach.

If it were up to me I'd switch `-Wextra` to `-Wimplicit-fallthrough=2` or make the regex of level 3 less project-specific. It doesn't make much sense to have this conditional to a leading or trailing single dash adjacent to the "fallthrough" word.

But well, I won't fight the author of the warning over it :)

marxin commented:
> IMO the formats used in those warnings levels are quite project-specific, so it looks like the author upstreamed their own project's convention together with the warning... but anyway, the `__attribute__(fallthrough)` is the most reliable one, so what we merge is the right approach.

Yes, I guess it was based on distribution rebuild. I must agree that the attribute is better (and with `[[fallthrough]]` even more portable).

akien-mga commented:
Cherry-picked e4a9616 for 3.1.1.
I'll wait for 3.1.1 to be released before cherry-picking the other main commit, just to be safe.

akien-mga commented:
Cherry-picked for 3.1.2.
